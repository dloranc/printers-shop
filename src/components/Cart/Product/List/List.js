import React, { Component } from 'react'
import { connect } from 'react-redux';

export class CartProductList extends Component {
  showProductList = () => {
    return (
      <>
        {this.props.products.map(product => (
          <div key={product.id}>{product.name}</div>
        ))}

        <div>Total price: {this.showTotalPrice()}</div>
      </>
    )
  }

  showCartIsEmpty = () => {
    return <p>Your cart is empty!</p>;
  }

  showTotalPrice = () => {
    return this.props.products.reduce((sum, product) => sum + product.amount * product.price, 0);
  }

  render() {
    return this.props.products.length > 0 ? this.showProductList() : this.showCartIsEmpty()
  }
}

const mapStateToProps = (state) => {
  return {
      products: state.cart,
  }
};

export default connect(mapStateToProps)(CartProductList);
