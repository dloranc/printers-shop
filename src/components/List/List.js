import React, { PureComponent } from 'react';
import Plant from './Plant/Plant';

export class List extends PureComponent {
  state = {
    showFruits: true,
    name: '',
    list: [
      { name: 'apples', type: 'fruit'},
      { name: 'bananas', type: 'fruit'},
      { name: 'potatoes', type: 'vegetable'},
      { name: 'strawberries', type: 'fruit'}
    ]
  }

  toggleFruits = () => {
    this.setState({
      showFruits: !this.state.showFruits
    });
  }

  handleClick = name => {
    this.setState({
      item: name
    });
  }

  handleRemove = name => {
    this.setState({
      list: [...this.state.list.filter(item => name !== item.name)]
    });
  }

  render() {
    return (
      <>
        {
          this.state.item
            ? `You like ${this.state.item}!`
            : 'Click on plant you like.'
        }

        <button
          id="toggle-fruits"
          onClick={this.toggleFruits}
        >
          {this.toggleFruits ? 'Hide fruits' : 'Show fruits'}
        </button>

        <div className="list">
          {this.state.list.length > 0 && this.state.list.map(
            item => {
              if (this.state.showFruits) {
                return this._renderPlant(item);
              }

              return item.type !== 'fruit'
                ? this._renderPlant(item)
                : null;
            }
          )}

          {this.state.list.length === 0 && <p>The list is empty.</p>}
        </div>
      </>
    );
  }

  _renderPlant = item => {
    return (
      <Plant
        name={item.name}
        onClick={this.handleClick}
        onRemove={this.handleRemove}
        key={item.name}
      />
    );
  }
}

export default List;
