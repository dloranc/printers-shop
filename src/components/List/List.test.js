import List from './List';
import toJson from 'enzyme-to-json';

// eslint-disable-next-line no-undef
const setup = buildSetup(List);

let wrapper = null;

describe('List component', () => {
  beforeEach(() => {
    wrapper = setup().wrapper;
  });

  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot after click', () => {
    wrapper.find('#toggle-fruits').simulate('click');

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render all types of plants', () => {
    expect(wrapper.text()).toContain('apples');
    expect(wrapper.text()).toContain('potatoes');
  });

  it('should have first element to be "apples"', () => {
    expect(wrapper.find('.list div:first-child').text()).toEqual('apples');
  });

  it('should render only vegetables when toggle button has been clicked',
    () => {
      wrapper.find('#toggle-fruits').simulate('click');

      expect(wrapper.text()).not.toContain('apples');
      expect(wrapper.text()).toContain('potatoes');
    }
  );
});
