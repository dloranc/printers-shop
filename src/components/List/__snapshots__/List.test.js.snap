// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should match snapshot 1`] = `
<Fragment>
  Click on plant you like.
  <button
    id="toggle-fruits"
    onClick={[Function]}
  >
    Hide fruits
  </button>
  <div
    className="list"
  >
    <Plant
      key="apples"
      name="apples"
      onClick={[Function]}
    />
    <Plant
      key="bananas"
      name="bananas"
      onClick={[Function]}
    />
    <Plant
      key="potatoes"
      name="potatoes"
      onClick={[Function]}
    />
    <Plant
      key="strawberries"
      name="strawberries"
      onClick={[Function]}
    />
  </div>
</Fragment>
`;

exports[`List component should match snapshot after click 1`] = `
<Fragment>
  Click on plant you like.
  <button
    id="toggle-fruits"
    onClick={[Function]}
  >
    Hide fruits
  </button>
  <div
    className="list"
  >
    <Plant
      key="potatoes"
      name="potatoes"
      onClick={[Function]}
    />
  </div>
</Fragment>
`;
